### PR TITLE
Storage Browser Default Auth

### PR DESCRIPTION
**Pull request by ashika112**
#### Description of changes
- Creates Amplify Auth adapter _[This is expected to be moved to UI. Including this atm to make testing easier.]_
- Creates ListPaths which is equivalent to ListLocations
- Updates type for ListPaths

#### TODO
- Pagination + Memoization
- Aligning behavior on user group edge case

#### Description of how you validated changes
- Unit test
- Manual testing using sample app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/__tests__/internals/amplifyAuthAdapter/createAmplifyAuthConfigAdapter.test.ts
Comment (ashika112) [inline packages/storage/__tests__/internals/amplifyAuthAdapter/createAmplifyAuthConfigAdapter.test.ts:1]: most likely will be moved to UI

File: packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyAuthConfigAdapter.ts
Comment (ashika112) [inline packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyAuthConfigAdapter.ts:1]: Adapter will most likely be moved to UI

File: packages/storage/src/internals/types/credentials.ts [hunk 1]
```diff
@@ -107,3 +107,37 @@ export interface LocationAccess extends CredentialsLocation {
...
+ */
+export interface PathAccess {
+	/** The Amplify backend mandates that all paths conclude with '/*',
+	 * which means the only applicable type in this context is 'PREFIX'. */
+	type: 'PREFIX';
+	permission: StorageAccess[];
+	scope: {
+		bucketName: string;
+		path: string;
+	};
```
Comment (AllanZhengYP) [inline packages/storage/src/internals/types/credentials.ts:122]: Subject to API review tho but it's better to align the scope here with the scope from other API: https://github.com/ashika112/amplify-js/blob/9cbc71d8a78ce8944db9642d7222efaec55f99e1/packages/storage/src/internals/types/credentials.ts#L81

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	identityId?: string;
+	userGroup?: string;
+}): PathAccess[] => {
+	const locations: PathAccess[] = [];
+
+	for (const [, bucketInfo] of Object.entries(buckets)) {
+		const { bucketName, paths } = bucketInfo;
+		if (!paths) {
+			// Todo: Verify behavior
+			return locations;
```
Comment (AllanZhengYP) [inline packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:50]: Does this mean default bucket? Should it be `continue`?

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+						path: path.replace(ENTITY_IDENTITY_URL, identityId),
+					},
+				});
+			}
+			const location = {
+				type: 'PREFIX',
+				...resolvePermissions(accessRules, tokens, userGroup),
+				scope: { bucketName, path },
+			};
+			if (location.permission) locations.push(location as PathAccess);
```
Comment (AllanZhengYP) [inline packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:68]: Sanity check, what are the cases when a path is omitted from `locations`? I'm wondering if we should fail-fast in these scenarios.

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	const locations: PathAccess[] = [];
+
+	for (const [, bucketInfo] of Object.entries(buckets)) {
+		const { bucketName, paths } = bucketInfo;
+		if (!paths) {
+			// Todo: Verify behavior
+			return locations;
+		}
+		for (const [path, accessRules] of Object.entries(paths)) {
+			if (tokens && identityId && path.includes(ENTITY_IDENTITY_URL)) {
```
Comment (AllanZhengYP) [inline packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:53]: Nit: this can be moved to a function similar to `resolvePermissions`

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts (hunk 3 above)
Comment (AllanZhengYP) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:68]: Nit: Looking at our doc it seems backend also support `Functions` access type. It make sense to not support it here, but just make sure we document it.

File: packages/storage/src/internals/types/common.ts [hunk 5]
```diff
@@ -20,3 +20,8 @@ export type Privilege = 'Default' | 'Minimal';
  * @internal
  */
 export type PrefixType = 'Object';
+
+/**
+ * @internal
+ */
+export type StorageAccess = 'read' | 'get' | 'list' | 'write' | 'delete';
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/internals/types/common.ts:27]: Yes verified

File: packages/storage/src/internals/types/credentials.ts (hunk 1 above)
Comment (ashika112) [reply packages/storage/src/internals/types/credentials.ts:122]: @AllanZhengYP I check in with Caleb on this, he proposed a different change in interface for both managedAuth and AmplifyAuth locations. Will raise a sep PR when we make that update everywhere

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts (hunk 2 above)
Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:50]: This is for when there is no paths. So locations will be [] at that point.

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts (hunk 3 above)
Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:68]: > what are the cases when a path is omitted from locations?
- There is logic in `resolvePermissions` on what permission to pick for what use cases like authenticated, unAuthenticated, user group, entity etc. Whereever we could fail fast that is done. This logic here is more santizing the array after resolvePermission.

Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:68]: > Nit: Looking at our doc it seems backend also support Functions access type. It make sense to not support it here, but just make sure we document it.

This will result in empty `paths` it actually would not even make it to `buckets` but for both of these scenarios we fail fast so it would not even get to this point

File: packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyListLocationsHandler.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+	return async function listLocations() {
+		if (!buckets) {
+			return { locations: [] };
+		}
+
+		const { tokens, identityId } = await fetchAuthSession();
+		const userGroups = tokens?.accessToken.payload['cognito:groups'];
+
+		const locations = generateLocationsFromPaths({
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyListLocationsHandler.ts:21]: Yes i have been trying to find the right naming too. I will updated this

File: packages/storage/src/internals/types/credentials.ts (hunk 1 above)
Comment (ashika112) [reply packages/storage/src/internals/types/credentials.ts:122]: Interface we want to move to should be along the lines of:
```
export interface LocationData {
  bucket: string;
  permission: Permission;
  prefix: string;
  type: LocationType;
}
```

Comment (calebpollman) [reply packages/storage/src/internals/types/credentials.ts:122]: Let's update the data output here to reflect @ashika112 above comment

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts (hunk 4 above)
Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:53]: This is for fail fast but we can create a util for this for sure.

File: packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts (hunk 2 above)
Comment (ashika112) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts:50]: I see what you meant. My bad. Updated test and logic.

File: packages/storage/src/internals/types/credentials.ts (hunk 1 above)
Comment (ashika112) [reply packages/storage/src/internals/types/credentials.ts:122]: @AllanZhengYP @calebpollman Updated the above mentioned structure just within Path context for this PR. I will open another PR to combine these type into one as LocationAccess.

File: packages/storage/src/internals/types/credentials.ts [hunk 7]
```diff
@@ -107,3 +107,36 @@ export interface LocationAccess extends CredentialsLocation {
 	 */
 	readonly type: LocationType;
 }
+
+/**
+ * @internal
+ */
+export interface PathAccess {
```
Comment (ashika112) [inline packages/storage/src/internals/types/credentials.ts:114]: Keeping this separate for now. In a separate PR will align this and managed Auth Location access into one.

File: packages/core/src/singleton/AmplifyOutputs/types.ts [hunk 8]
```diff
@@ -50,6 +50,8 @@ export interface AmplifyOutputsStorageBucketProperties {
 	bucket_name: string;
 	/** Region for the bucket */
 	aws_region: string;
+	/** Paths to object with access permissions */
+	paths?: Record<string, Record<string, string[] | undefined>>;
```
Comment (AllanZhengYP) [inline packages/core/src/singleton/AmplifyOutputs/types.ts:54]: The in-line type is a bit hard the follow, what about using a interface with property names:
```typescript
type Permission = string;

interface StorageBucketPaths {
  [pathPrefix: string]: {
    [accessType: string]: Permission[] | undefined;
  }
}
```

**Review by AllanZhengYP — COMMENTED** [review]
Looks good to me! Only have 1 nit comment.

File: packages/core/src/singleton/AmplifyOutputs/types.ts (hunk 8 above)
Comment (ashika112) [reply packages/core/src/singleton/AmplifyOutputs/types.ts:54]: Our ESLint does not allow for index signature. But i think we can create alias and align if we want. I can look into this in my follow up PR for pagination if there are no other comments :)

**Review by AllanZhengYP — APPROVED** [review]

File: packages/storage/src/internals/types/common.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/storage/src/internals/types/common.ts:27]: We still need to align on whether we are using this string union or the AG `permission` in the component

File: packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyListLocationsHandler.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { Amplify, fetchAuthSession } from '@aws-amplify/core';
+
+import { ListPaths } from '../types/credentials';
+
+import { generateLocationsFromPaths } from './generateLocationsFromPaths';
+
+export const createAmplifyListLocationsHandler = (): ListPaths => {
+	const { buckets } = Amplify.getConfig().Storage!.S3!;
```
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/storage/src/internals/amplifyAuthConfigAdapter/createAmplifyListLocationsHandler.ts:11]: If we move this lookup inside of `listLocations` think we can just export `listLocations` without the `createAmplifyListLocationsHandler` HOF?

File: packages/storage/src/internals/types/credentials.ts [hunk 10]
```diff
@@ -107,3 +107,36 @@ export interface LocationAccess extends CredentialsLocation {
...
+ */
+export interface ListPathsOutput {
+	locations: PathAccess[];
+	nextToken?: string;
+}
+
+/**
+ * @internal
+ */
+export type ListPaths = (input?: ListPathsInput) => Promise<ListPathsOutput>;
```
[the remark this answers is not in the document]
Comment (ashika112) [reply packages/storage/src/internals/types/credentials.ts:142]: Yes working with Hui on this. I dont think we need to block this PR for it. Since depending on how we move the adapter into UI we can have accommodate type change accordingly

**Review by calebpollman — APPROVED** [review]
